android/src/main/java/com/tailscale/ipn/NetworkChangeCallback.kt

```diff
@@ -2,55 +2,85 @@
 // SPDX-License-Identifier: BSD-3-Clause
 package com.tailscale.ipn
 
-import android.content.Context
 import android.net.ConnectivityManager
 import android.net.LinkProperties
 import android.net.Network
 import android.net.NetworkCapabilities
-import android.net.NetworkRequest
 import android.util.Log
 import libtailscale.Libtailscale
 import java.net.InetAddress
-import java.net.NetworkInterface
 
 object NetworkChangeCallback {
 
+  // Track the current active network
+  @Volatile private var activeNetwork: Network? = null
+  private val TAG = "NetworkChangeCallback"
+
+  // Create a lock object for synchronizing access to activeNetwork
+  private val activeNetworkLock = Any()
+
   // requestNetwork attempts to find the best network that matches the passed NetworkRequest. It is
   // possible that this might return an unusuable network, eg a captive portal.
   fun monitorDnsChanges(connectivityManager: ConnectivityManager, dns: DnsConfig) {
-    val networkConnectivityRequest =
-        NetworkRequest.Builder()
-            .addCapability(NetworkCapabilities.NET_CAPABILITY_INTERNET)
-            .addCapability(NetworkCapabilities.NET_CAPABILITY_NOT_VPN)
-            .build()
-
-    connectivityManager.registerNetworkCallback(
-        networkConnectivityRequest,
+
+    connectivityManager.registerDefaultNetworkCallback(
         object : ConnectivityManager.NetworkCallback() {
           override fun onAvailable(network: Network) {
-            super.onAvailable(network)
+            val networkCapabilities = connectivityManager.getNetworkCapabilities(network)
+            if (networkCapabilities != null &&
+                networkCapabilities.hasCapability(NetworkCapabilities.NET_CAPABILITY_NOT_VPN) &&
+                networkCapabilities.hasCapability(NetworkCapabilities.NET_CAPABILITY_INTERNET)) {
+              // Only use if the network in the callback is the current active network. Network
+              // updates are not guaranteed to come in order.
+              synchronized(activeNetworkLock) {
+                if (connectivityManager.activeNetwork == network) {
```

Comment:
Is this required now that we're only getting notified about the system's default network? The docs aren't super clear I think...

Reply:
iiuc it is not necessary